extensions/amp-analytics/0.1/vendors/snowplow_v2.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {jsonLiteral} from '../../../../src/json';
+
+const SNOWPLOW_V2_CONFIG = jsonLiteral({
+  'linkers': {
+    'linker': {
+      'ids': {
+        '_sp_duid': '${duid}',
+      },
+    },
+  },
+  'cookies': {
+    'enabled': true,
+    'cookieMaxAge': 1800,
+    '_sp_duid': {
+      'value':
+        '$IF($IF(COOKIE(_sp_duid), COOKIE(_sp_duid), LINKER_PARAM(linker,_sp_duid)), $IF(COOKIE(_sp_duid), COOKIE(_sp_duid), LINKER_PARAM(linker,_sp_duid)), $SUBSTR(QUERY_PARAM(_sp),0,36))',
+    },
+  },
+  'vars': {
+    'aaVersion': 'amp-1.0.0',
+    'ampVistorId': 'CLIENT_ID(_sp_id)',
+    'nullString': 'null',
+    'customEventTemplate': [
+      '{',
+      '"schema":"iglu:com.snowplowanalytics.snowplow/unstruct_event/jsonschema/1-0-0",',
+      '"data":{',
+      '"schema":"iglu:${customEventSchemaVendor}/${customEventSchemaName}/jsonschema/${customEventSchemaVersion}",',
+      '"data":{',
+      '${customEventSchemaData}',
+      '}',
+      '}',
+      '}',
+    ].join(''),
+    'contextsHead':
+      '{"schema":"iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0","data":[{"schema":"iglu:com.snowplowanalytics.amp/amp_id/jsonschema/1-0-0","data":{"ampClientId":"${ampVistorId}", "domainUserId": "COOKIE(_sp_duid)", "userId": "${userId}"}},${customContexts}',
```

Comment:
You may have already tried it, but I feel like you could avoid the leading/trailing comma issue with something like `$REPLACE($DEFAULT(customContexts, ''), '^,*(.+?),*$', '$1,')` which should work with leading commas, trailing commas, neither commas, both commas, empty string, and no value, only adding the comma to the end when appropriate. This might simplify quite a bit of the context handling? (you will undoubtedly have more testcases than I do)

Reply:
I hadn't, but I had tried quite a few things in a similar vein and just attempted this now - I haven't managed to get anything to work.

It might be that I've just not found how to use it properly, but it seems that something about these macros malforms the data - either it won't recognise the macro, or it won't substitute the variable correctly and will corrupt the request. 

I'd be happy to include something like this if we can get it to work.

Reply:
```
'{"schema":"iglu:com.snowplowanalytics.snowplow/contexts/jsonschema/1-0-0","data":[{"schema":"iglu:com.snowplowanalytics.amp/amp_id/jsonschema/1-0-0","data":{"ampClientId":"${ampVistorId}", "domainUserId": "COOKIE(_sp_duid)", "userId": "${userId}"}},$REPLACE(`${customContexts}`, `^,*(.+?),*$`, `$1,`)',
```
This _seems_ to work now. The magic is apparently the backticks which slightly changes the evaluation of the macros I think?

It breaks the example a bit, you then need to change `clientId(_ga)` to `CLIENT_ID(_ga)` but then it seems to work OK, I think?

Reply:
It does indeed work, thanks, have added this change.